### ISSUE TITLE
False positive error on L057 for project name with backticks

### ISSUE THREAD

**Issue by bbrewington** [sqlfluff/sqlfluff#3401]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

File "dummy_sql.sql":

```sql
select DUMMY_COL
from `project-name-with-dash.dataset_name.table_name`
```

Run sqlfluff in Windows Command Prompt:

```
> sqlfluff lint dummy_sql.sql --dialect bigquery
```

Result:
![2022-05-27 10_19_10-Window](https://user-images.githubusercontent.com/10573749/170717954-99028acf-d693-4a3d-851b-24cc2e387740.png)

### Expected Behaviour

Expect no linting errors from above query

Per Google Cloud's docs on "Table Path syntax" ([link](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#table_path)), a project name can be "quoted" via backticks if it contains special characters (in example above, a dash - in Google Cloud's example, a colon):

> If the project name contains special characters such as colons, quote the name in backticks ` (example: `google.com:my_project`).

### Observed Behaviour

See screenshot in "What Happened" section above

### How to reproduce

Reprex provided in "What Happened" section above

### Dialect

bigquery

### Version

sqlfluff, version 0.13.2
Python 3.9.6

### Configuration

default, don't have .sqlfluff or .sqlfluffignore

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by bbrewington** [comment]
Not very familiar with the Python code that's powering this behavior, but glad to look for it if someone points me in the right direction

**Comment by bbrewington** [comment]
I was able to get around this by adding this exception to .sqlfluff, but I think for dialect bigquery, given that they allow a particular set of special characters in table paths, it should be robustly accounted for in default behavior

```
[sqlfluff:rules:L057]
additional_allowed_characters = "-"
```

**Comment by tunetheweb** [comment]
The rule already handles backticks correctly forn the BigQuery dialect, but the error message is a little confusing - someone else raises this recently.

The issue is not the backticks, but the dashes. However the rule highlights the beginning of the identifier, which made you (not unreasonably to be fair!) think it was the backticks.

You can configure the rule with `additional_allowed_characters` to also allow dashes, on top of the allowed `[A-Za-z0-9_]` characters. Maybe that should be a default?

**Comment by tunetheweb** [comment]
> I was able to get around this by adding this exception to .sqlfluff, but I think for dialect bigquery, given that they allow a particular set of special characters in table paths, it should be robustly accounted for in default behavior

Our messages crossed there!

The aim of the rule (and of many linting rules!) is to have opinionated rules about what you _should_ use, not what you _can_ use. Technically `ac£54` is also a valid table name if enclosed in back ticks but I'd hope you'd agree that's not that useful a table name, nor best practice.

I can accept that dashes could be added to the default allowed set of characters. As I say you are not the first to raise this, though it should also be noted that dashes do have special rules in BigQuery (see https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical) and so cannot be used in a similar way as alphanumeric characters, perhaps highlighting that they should not be the default and people should consciously confige the rule to allow these (which, as you note, you can do).

**Comment by bbrewington** [comment]
Definitely tracking with what you're saying, and looks like typical practice is to just build in the exceptions your organization uses as conventions (so by default it's stricter, and user can widen the space out close to all that GCP allows)

I just noticed the info in Rule L031 for BigQuery dialect about complexity of determining project vs. dataset name: https://github.com/sqlfluff/sqlfluff/blob/0ae7b780da88b6a6f4db471774c42ace952e8f7a/src/sqlfluff/rules/L031.py#L43-L46

Including this for reference...here's some examples of repos in the wild using bigquery dialect and adding L057 related stuff

- tenajima/kawaii-dbt 
    - Excludes rule L057 altogether: https://github.com/tenajima/kawaii-dbt/blob/9ec116e922dc9ab7299ae220b24d3efbd429d471/.sqlfluff#L3-L4
- HTTPArchive
    - Allow "-."
        - https://github.com/HTTPArchive/almanac.httparchive.org/blob/5a8015abf0619ae49bb5e30d65ae0ec7c7251622/sql/.sqlfluff#L122
        - https://github.com/HTTPArchive/data-pipeline/blob/80a1e87901d75013c609b5d5daa0fdd5e3e2209a/.github/linters/.sqlfluff#L123
    - Allow ".-${}": https://github.com/HTTPArchive/bigquery/blob/7f866ccbeb583e42552ed2f0e72e9f6f417b3a8f/sql/.sqlfluff#L108

**Comment by bbrewington** [comment]
Closing this, since it seems to be expected behavior

**Comment by tunetheweb** [comment]
The HTTP Archive ones are mine btw 😁

Yeah it is expected behaviour. I’m on the fence myself whether we should add dashes.
